rosette/rbl/rosette/tests/run-tests.ros

```diff
@@ -19,15 +19,49 @@
 ;; Form (+ 1 2) = 4 failed (produced 3).
 ;; NIL
 
-(define failures 0)
-(defProc (incf-failures)
-  (set! failures (+ failures 1))
-  failures)
+;; Make an actor that emulates a counter. It can be incremented, decremented,
+;; and reset to zero.
+(defOprn increment)
+(defOprn decrement)
+(defOprn reset)
+(defOprn value)
 
-(define successes 0)
-(defProc (incf-successes)
-  (set! successes (+ successes 1))
-  successes)
+(defActor Counter (slots& value 0)
+  (method (increment & delta)
+     (let* [[delta (if (null? delta) 1 (head delta))]
+            [new-value (+ delta value)]]
+       (update value new-value)
+       new-value))
+  (method (decrement & delta)
+     (let* [[delta (if (null? delta) 1 (head delta))]
+            [new-value (- delta value)]]
+       (update value new-value)
+       new-value))
+  (method (reset)
+     (let []
+       (update value 0)
+       0))
+  (pure (value) value))
+
+;; Counters for logical sets of tests.
+(define failures (new Counter))
+(define successes (new Counter))
+
+;; Counters for total test success and failure counts.
+(define total-failures (new Counter))
+(define total-successes (new Counter))
+
+;; Call this at the end of a logical group of test. It will print out if there
+;; were any failures in that group and accumulate total successes and failure
+;; counts before resetting the counters.
+(defProc (record-and-reset)
+  (if (> (value failures) 0)
+      (display "Successes: " (value successes) ". Failures: " (value failures) ".\n")
+      (display "All sub-tests passed.\n"))
+  (increment total-successes (value successes))
+  (increment total-failures (value failures))
+  (reset successes)
+  (reset failures))
```

Comment:
*** warning: more than one result may be returned from a block expression

While in practice this proc executes sequentially, in theory it can call (reset) before (increment) which may lead to a confusing output. Thus the whole thing should be wrapped with a (seq ...).

Reply:
Ah, you're right. I learned this later than I wrote this.